Framework/Kernel/src/StringTokenizer.cpp

```diff
@@ -0,0 +1,171 @@
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   if (first != second)
+                     output.emplace_back(first, second);
+                 });
+  return output;
+}
+
+// split trimming whitespace and keeping empty tokens.
+std::vector<std::string> split2(const std::string &str,
+                                const std::string &delims) {
+  std::vector<std::string> output;
+  for_each_token(str.cbegin(), str.cend(), delims.cbegin(), delims.cend(),
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   output.emplace_back(first, second);
+                   trim(output.back());
+                 });
+  return output;
+}
+
+// split trimming whitespace and ignoring empty tokens.
+std::vector<std::string> split3(const std::string &str,
+                                const std::string &delims) {
+  std::vector<std::string> output;
+  for_each_token(str.cbegin(), str.cend(), delims.cbegin(), delims.cend(),
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   if (first != second) {
+                     output.emplace_back(first, second);
+                     trim(output.back());
+                     if (output.back().empty())
+                       output.pop_back();
+                   }
+                 });
+  return output;
+}
+}
+
+/**
+ * Constructor requiring a string to tokenize and a string of separators.
+ * @param str Input string to be separated into tokens.
+ * @param separators List of characters used to separate the input string.
+ * @param options  tokenizer settings. The number can be found using the
+ * StringTokenizer::Options enum
+ * @throw Throws std::runtime_error if options > 7.
+ * @return a const reference to the index'th token.
+ */
+Mantid::Kernel::StringTokenizer::StringTokenizer(const std::string &str,
+                                                 const std::string &separators,
+                                                 unsigned options) {
+  // check options variable is in the range 0-7.
+  if (options > 7)
+    throw std::runtime_error(
+        "Invalid option passed to Mantid::Kernel::StringTokenizer:" +
+        std::to_string(options));
+  // if str is empty, then there is no workt to do. exit early.
+  if (str.empty())
+    return;
+
+  // see comments above for the different options split0,split1,split2 and
+  // split3 implement.
+  // cases 0-3 will check for a separator in the last place and insert an empty
+  // token at the end.
+  // cases 4-7 will not check and ignore a potential empty token at the end.
+  switch (options) {
+  case 0:
+    m_tokens = split0(str, separators);
+    emptyFinalToken(str, separators, m_tokens);
+    break;
+  case 1:
```

Comment:
I think it may improve readability to change the integers into `TOK_TRIM`, `TOK_IGNORE_EMPTY | TOK_TRIM` and so on. Does this have any impact on performance?